v-next/hardhat/src/internal/cli/telemetry/sentry/reporter.ts

```diff
@@ -0,0 +1,120 @@
+
+// TODO: replace with PROD version
+// export const SENTRY_DSN =
+//   "https://38ba58bb85fa409e9bb7f50d2c419bc2@o385026.ingest.sentry.io/522486955555555555";
+export const SENTRY_DSN =
+  "https://d578a176729662a28e7a8da268d36912@o385026.ingest.us.sentry.io/4507685793103872"; // DEV
+
+export async function sendErrorTelemetry(error: Error): Promise<boolean> {
+  const instance = await Reporter.getInstance();
+  return instance.reportError(error);
+}
+
+// ATTENTION: this function is exported for testing, do not directly use it in production
+export function _testResetReporter(): void {
+  Reporter.deleteInstance();
+}
+
+class Reporter {
+  // GENERAL EXPLANATION:
+  // 1) The 'reportError' function collects the error and passes it to our custom Sentry transporter.
+  // 2) The custom transporter receives the JavaScript error serialized by Sentry.
+  // 3) This serialized error is then passed to a detached subprocess, which anonymizes all the information before sending it to Sentry.
+
+  static #instance: Reporter | undefined;
+  readonly #telemetryEnabled: boolean;
+
+  private constructor(telemetryAllowed: boolean) {
+    this.#telemetryEnabled = telemetryAllowed;
+  }
+
+  public static async getInstance(): Promise<Reporter> {
+    if (this.#instance !== undefined) {
+      return this.#instance;
+    }
+
+    const telemetryAllowed = await isTelemetryAllowed();
+    this.#instance = new Reporter(telemetryAllowed);
+
+    const { Integrations, init, setExtra } = await import("@sentry/node");
+
+    const linkedErrorsIntegration = new Integrations.LinkedErrors({
+      key: "parent",
+    });
+
+    const transport = await getSubprocessTransport();
+
+    init({
+      dsn: SENTRY_DSN,
+      transport,
+      integrations: () => [linkedErrorsIntegration],
+    });
+
+    setExtra("nodeVersion", process.version);
+    setExtra("hardhatVersion", await getHardhatVersion());
+
+    return this.#instance;
+  }
+
+  public static deleteInstance(): void {
+    // ATTENTION: only for testing
+    this.#instance = undefined;
+  }
+
+  public async reportError(
+    error: Error,
+    verbose: boolean = false,
+    configPath: string = "",
+  ): Promise<boolean> {
+    if (!(await this.canSendTelemetry(error))) {
+      return false;
+    }
+
+    const { captureException, setExtra } = await import("@sentry/node");
+
+    setExtra("verbose", verbose);
+    setExtra("configPath", configPath);
+
+    captureException(error);
+
+    // TODO: where to run it?
+    // await this.#sentry.flush(1000);
+
+    return true;
+  }
+
+  private async canSendTelemetry(error: Error): Promise<boolean> {
+    if (!this.#telemetryEnabled) {
+      return false;
+    }
+
+    if (
+      HardhatError.isHardhatError(error) &&
+      !error.descriptor.shouldBeReported
+    ) {
+      return false;
+    }
+
+    if (HardhatPluginError.isHardhatPluginError(error)) {
+      // Don't log errors from third-party plugins
+      return false;
+    }
+
```

Comment:
We should probably be a bit smarter here, and NOT send the errors that we know aren't ours (i.e. like a blocklist). For example, errors sent by ethers.js.

This can be a separate task, but we should do it.

Reply:
What I mean is checking the stack trace, and if it comes from a well-known file/package that we don't want to report, we just return false.

Example:

```
Error: ERROR processing <user-file>
Error: invalid address or index (argument="addressOrIndex", value="<SignerWithAddress xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx>", code=INVALID_ARGUMENT, version=provide...
  File "node_modules/@ethersproject/logger/src.ts/index.ts", line 269, col 28, in Logger.makeError
    const error: any = new Error(message);
  File "node_modules/@ethersproject/logger/src.ts/index.ts", line 281, col 20, in Logger.throwError
    throw this.makeError(message, code, params);
  File "node_modules/@ethersproject/logger/src.ts/index.ts", line 285, col 21, in Logger.throwArgumentError
    return this.throwError(message, Logger.errors.INVALID_ARGUMENT, {
  File "node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts", line 183, col 20, in new JsonRpcSigner
    logger.throwArgumentError("invalid address or index", "addressOrIndex", addressOrIndex);
  File "node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts", line 489, col 16, in EthersProviderWrapper.JsonRpcProvider.getSigner
    return new JsonRpcSigner(_constructorGuard, this, addressOrIndex);
  File "node_modules/@nomiclabs/hardhat-ethers/src/internal/helpers.ts", line 62, col 38, in getSigner
    const signer = hre.ethers.provider.getSigner(address);
  File "<user-file>", line 95, col 5, in processTicksAndRejections
  File "<user-file>", line 275, col 32, in async createAndInitialize_UniV2_PoolForPairs
  File "<user-file>", line 411, col 13, in async processPairsWithParams
```

Reply:
[Done here](https://github.com/NomicFoundation/hardhat/pull/5570/commits/bed832cd1836c4fe16d1f965686e6e826884a039). But let's discuss if we want to use a different approach